updated vault to fetch token from file

murphymj25 opened the pull request:
#523 Updated vault to fetch token from file at start and at refresh interval.

pschultz commented on cert/vault_client.go:
```diff
@@ -23,11 +24,14 @@ type vaultClient struct {
 
 	if c.client != nil {
+		if vaultTokenFromFile == "true" {
+			c.client.SetToken(getVaultTokenFromFile(vaultTokenPath))
```
The token should only be modified if the file has been read successfully and the token isn't empty after stripping whitespace. Otherwise we should continue to use the existing token I think.

Also, when configuring the client with the initial token we support wrapped tokens, but here we don't. We can probably re-use the logic below, but it's a bit tricky.

We can reasonably expect that a wrapped token only works once, since that's the whole point of wrapping. In other words, the token stored in the file may be useless after the inital setup, so we can't blindly re-set the token here. We must do that only if the file contents changed, and do the unwrapping step again.

murphymj25 replied:
For my use case, i trust that the token in the file is always a valid token, so i would want to grab the valid token every time fabio does a new check to vault, but i can see where that might be a challenge. I'll add some logic to only set the token if it has changed in the file (the current token no longer matches the token in the file), and account for if the file is blank or missing.

pschultz commented on fabio.properties:
```diff
@@ -137,6 +137,13 @@
+# Get vault token from file
+#
+# The 'tokenfromfile' option enables fetching the vault token from a file on the filesystem.  If
+# 'tokenfromfile' is set to "true", provide the path in the 'tokenpath' option.
```
Capitalize Vault, please, and also update the docs for the Vault KV source a few paragraphs above, where it says "The token must be provided in the VAULT_TOKEN environment variable."

murphymj25 replied:
Will do

pschultz commented on cert/vault_client.go:
```diff
@@ -132,3 +137,14 @@ func (c *vaultClient) keepTokenAlive() {
+		log.Printf("[WARN] vault: Failed to get Token From File: %s", err)
+	}
+	str := string(b)
+	str = strings.TrimSuffix(str, "\n")
```
Any reason to not call `strings.TrimSpace` here? Some config management tools are finicky about whitespace in their template engines, for instance. Copy-pasting can also easily add whitespace by accident.

murphymj25 replied:
I'll update to TrimSpace.

pschultz commented on cert/vault_client.go:
```diff
@@ -132,3 +137,14 @@ func (c *vaultClient) keepTokenAlive() {
+func getVaultTokenFromFile(c string) string {
+	b, err := ioutil.ReadFile(c) // just pass the file name
+	if err != nil {
+		log.Printf("[WARN] vault: Failed to get Token From File: %s", err)
```
nit: `Token From File` should be lowercase. In the debug message below it should say "fetched", not "fetch".

murphymj25 replied:
Will do

murphymj25 commented on config/config.go:
```diff
@@ -22,14 +22,16 @@ type Config struct {
+	Refresh            time.Duration
+	Header             http.Header
+	VaultTokenFromFile string
+	VaultTokenPath     string
```
I think that makes sense.  I'll get this updated.

murphymj25 commented on config/load.go:
```diff
@@ -531,6 +531,10 @@ func parseCertSource(cfg map[string]string) (c CertSource, err error) {
+		case "tokenfromfile":
+			c.VaultTokenFromFile = v
+		case "tokenpath":
+			c.VaultTokenPath = v
```
Will do.

murphymj25 commented:
Thanks for the feedback!  I'll look to incorporate your suggestions.  Do you have a link that gives an overview of wrapped tokens?  That may help my understanding, but I think i get the general idea based on your comments.

pschultz commented:
Wrapping is explained in detail here: https://www.vaultproject.io/docs/concepts/response-wrapping.html

In short: it may not be desirable to pass an actual Vault token (or any other secret, really) around, especially if it ends up in a file on disk.

Vault offers the possibility to wrap secrets to get around this. Instead of returning the secret directly, Vault creates a so-called wrapping token that can be used exactly once and stores the secret in that wrapping token's cubbyhole. You then give the wrapping token to fabio, and fabio trades it for the actual token. The corresponding CLI commands look something like this:

```txt
$ vault token create -policy fabio -wrap-ttl=5m
Key                              Value
---                              -----
wrapping_token:                  s.E0mxDIk9sZBTvDtZXe7ZMH3S
wrapping_token_ttl:              5m
[...]
```

```txt
$ vault token lookup s.E0mxDIk9sZBTvDtZXe7ZMH3S
Key                 Value
---                 -----
explicit_max_ttl    5m
num_uses            1
policies            [response-wrapping]
[...]
```

```txt
$ vault unwrap s.E0mxDIk9sZBTvDtZXe7ZMH3S
Key                  Value
---                  -----
token                s.Lcxkbr1PYZ5R2chAcALh0XMV
[...]
```

```txt
$ vault token lookup s.Lcxkbr1PYZ5R2chAcALh0XMV
Key                 Value
---                 -----
explicit_max_ttl    0s
num_uses            0
policies            [default fabio]
[...]
```

`unwrap` returns what `token create` would have returned if the `-wrap` flag hadn't been specified.

In this example the wrapping token is revoked after 5 minutes, or after the first unwrap request, whichever happens first. This makes it pretty safe to store it on disk. Even if someone gets their hands on it, it's worthless, assuming they don't beat fabio to the punch.

The consequence is that it is not enough to compare the file content with client's token. If the token was wrapped, they *will* be different. You have to remember the file content separately after a reload and compare against *that* value.

murphymj25 commented:
Got it, i think i can put something together for that.  Thanks for the additional detail!

murphymj25 commented:
Just wanted to give a quick update.  Using wrapper tokens makes a lot of sense for my use case, so I'm working to get my test vault environment setup to support that.  Once i have that in place i should be able to make all of the proposed changes and be able to validate.

murphymj25 commented:
@pschultz I've updated the PR with your suggestions.  I also changed things a bit, you can now do 
vaultfetchtoken=env:[var] to pull from and environment variable and vaultfetchtoekn=file:[path] to pull it from a file.  I still have to test with a wrapped token, but the logic should be in place to support it.  I'll update this thread once i confirm wrapped tokens behave as expect.

murphymj25 commented:
I have tested with wrapped tokens, and everything looks to be working as expected.

murphymj25 commented on cert/vault_client.go:
```diff
@@ -23,11 +25,21 @@ type vaultClient struct {
+			if token != prevFetchedToken {
+				log.Printf("[DEBUG] Vault: token has changed, setting new token")
+				c.client.SetToken(token)
+				prevFetchedToken = token
```
I've tested this and when i add a new token to the file i see that it get's unwrapped and is used.  Perhaps i'm missing something

murphymj25 commented on cert/vault_client.go:
```diff
@@ -23,11 +25,21 @@ type vaultClient struct {
 var DefaultVaultClient = &vaultClient{}
 
-func (c *vaultClient) Get() (*api.Client, error) {
+var prevFetchedToken string
```
I forgot you can have multiples (we only use on in our env), i'll make these updates.

murphymj25 commented on fabio.properties:
```diff
@@ -114,8 +114,8 @@
-# variable. The token must be provided in the VAULT_TOKEN environment
-# variable.
+# variable. The token can be provided in the VAULT_TOKEN environment
+# variable, or provided by using the Vault fetch token option.
```
Sure thing!

murphymj25 commented on cert/vault_client.go:
```diff
@@ -23,11 +25,21 @@ type vaultClient struct {
+		if fetchVaultToken != "" {
+			token := strings.TrimSpace(getVaultToken(fetchVaultToken))
+			if token != prevFetchedToken {
+				log.Printf("[DEBUG] Vault: token has changed, setting new token")
```
Wil do!

murphymj25 commented on cert/vault_client.go:
```diff
@@ -39,16 +51,22 @@ func (c *vaultClient) Get() (*api.Client, error) {
 
+	if fetchVaultToken != "" {
+		token := strings.TrimSpace(getVaultToken(fetchVaultToken))
+		log.Printf("[DEBUG] Vault: fetching initial token")
```
Wil do!

murphymj25 commented on cert/vault_client.go:
```diff
@@ -132,3 +150,24 @@ func (c *vaultClient) keepTokenAlive() {
+
+func getVaultToken(c string) string {
+	c = strings.TrimSpace(c)
+	cArray := strings.Split(c, ":")
```
Will do!

murphymj25 replied:
I had to set the last argument to 2, but otherwise this seems to now be working.

murphymj25 commented on cert/vault_client.go:
```diff
@@ -132,3 +150,24 @@ func (c *vaultClient) keepTokenAlive() {
+	cArray := strings.Split(c, ":")
+	var token string
+	if cArray[0] == "file" {
+		b, err := ioutil.ReadFile(cArray[1]) // just pass the file name
```
Will do!

pschultz replied:
strings.SplitN does not replace a bounds check; it returns at most 2 substrings, not at least: https://play.golang.org/p/qkKS5Es2f9T

murphymj25 commented on fabio.properties:
```diff
@@ -137,6 +137,15 @@
+# The 'vaultfetchtoken' option enables fetching the vault token from a file on the filesystem or an environment
+# variable at the Vault refresh interval.  If fetching the token from a file the 'file:[path]' syntax should be used,
+# if fetching the token from an env variable, the 'env:[ENV]' syntax should be used.
+# in the 'tokenpath' option.
```
Will do!

murphymj25 replied:
I thought this is still needed as it give more detail on the vaultfetchtoken setting.

pschultz replied:
I was just referring to the last line here:

> in the 'tokenpath' option.

The rest of the paragraph is fine.

murphymj25 commented:
@pschultz Thanks for the feedback.  I'll address all of the items.  The unwrapping the token a 2nd time was a miss on my part; i'll get it updated.

murphymj25 commented:
@pschultz I have made all of the requested updates.  I'm just going to have a peer of mine review everything quick.  I should be able to push the updates next week.

murphymj25 commented:
I've updated the PR with the latest changes.

murphymj25 commented:
@pschultz Thanks, I'll take a look at adding a check.  When I tested I wan't getting a failure, if it's failing in the playground it would be good to address.

murphymj25 commented:
@pschultz I have added a check that should handle if something isn't set correctly.  Also cleaned up that line in properties, and ran a go fmt.

pschultz commented:
Sorry for the delay; I was on vacation.

I can't figure out how to break this anymore, so the PR looks good to me now! Thanks a lot.

There are, however, two unreachable log statements in vault_client.go (the log calls and return statements should be swapped):

```txt
$ go vet ./cert
# github.com/fabiolb/fabio/cert
cert/vault_client.go:184:4: unreachable code
cert/vault_client.go:192:4: unreachable code
```

@leprechau, do you want to give this a once-over?

murphymj25 commented:
Thanks @pschultz, I'll get those returns updated.

murphymj25 commented:
@pschultz I've updated the log statements.  Thanks!

murphymj25 commented:
Also, the "log.Printf("[WARN] vault: vaultfetchtoken not properly set")" was being sent every time there was an err, i've added a check so that log makes it more clear when something is not set correctly.

leprechau commented:
LGTM!  Thank you both for all the work.

pschultz commented:
I quess a manual squashed merge isn't picked up by Github.

Merged at 146ec88ea3d81a3f6a24f9d9e8819566fa2cb066. Thanks again!

murphymj25 commented:
@pschultz Thanks for all of the help with this!
